lib/msal-core/src/utils/RequestUtils.ts

```diff
@@ -129,11 +137,55 @@ export class RequestUtils {
      * @ignore
      *
      * generate unique state per request
-     * @param request
+     * @param userState User-provided state value
+     * @returns State string include library state and user state
+     */
+    static validateAndGenerateState(userState: string, interactionType: InteractionType): string {
+        return !StringUtils.isEmpty(userState) ? `${RequestUtils.generateLibraryState(interactionType)}${Constants.resourceDelimiter}${userState}` : RequestUtils.generateLibraryState(interactionType);
+    }
+
+    /**
+     * Generates the state value used by the library.
+     *
+     * @returns Base64 encoded string representing the state
+     */
+    static generateLibraryState(interactionType: InteractionType): string {
+        const stateObject: LibraryStateObject = {
+            id: CryptoUtils.createNewGuid(),
+            ts: TimeUtils.now(),
+            method: interactionType
+        };
+
+        const stateString = JSON.stringify(stateObject);
+
+        return CryptoUtils.base64Encode(stateString);
+    }
+
+    /**
+     * Decodes the state value into a StateObject
+     *
+     * @param state State value returned in the request
+     * @returns Parsed values from the encoded state value
      */
-    static validateAndGenerateState(state: string): string {
-        // append GUID to user set state  or set one for the user if null
-        return !StringUtils.isEmpty(state) ? CryptoUtils.createNewGuid() + "|" + state : CryptoUtils.createNewGuid();
+    static parseLibraryState(state: string): LibraryStateObject {
+        const libraryState = state.split(Constants.resourceDelimiter)[0];
+
+        if (CryptoUtils.isGuid(libraryState)) {
+            return {
+                id: libraryState,
+                ts: TimeUtils.now()
+            };
```

Comment:
In this scenario, I think we may want to have `method` default to redirect, to error on teh side of the response getting processed (as I believe redirect is the most likely scenario where this could happen).